### PR TITLE
Wire LedgerEndCache to relevant StorageBackend factories [DPP-704]

### PR DESCRIPTION
**Pull request by nmarton-da**
* Add ledgerEndCache dependency to relevant StorageBackend factory methods
* Wire ledgerEndCache to appropriate places
* REFACTORING: StorageBackendTests to instantiate TestBackend, wire ledgerEndCache update
* REFACTORING: RecoveringIndexerIntegrationSpec to simplify party existence, wire ledgerEndCache update
* REFACTORING: Remove StorageBackend super-trait, move StorageBackend DTO-s
* REFACTORING: Move common part of StorageBackendFactory to CommonStorageBackendFactory

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by nmarton-da** [comment]
FYI @rautenrieth-da @tudor-da about testing with storage backends with cache

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendProvider.scala [hunk 1]
```diff
@@ -5,43 +5,96 @@ package com.daml.platform.store.backend
...
   protected final def ingest(dbDtos: Vector[DbDto], connection: Connection): Unit = {
     def typeBoundIngest[T](ingestionStorageBackend: IngestionStorageBackend[T]): Unit =
       ingestionStorageBackend.insertBatch(connection, ingestionStorageBackend.batch(dbDtos))
-    typeBoundIngest(backendFactory.createIngestionStorageBackend)
+    typeBoundIngest(backend.ingestion)
+  }
+
+  protected final def updateLedgerEnd(ledgerEnd: LedgerEnd)(connection: Connection): Unit = {
+    backend.parameter.updateLedgerEnd(ledgerEnd)(connection)
+    updateLedgerEndCache(connection)
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendProvider.scala:32]: Out of interest. Is ledger end here and `head index` from `com.daml.platform.akkastreams.dispatcher.Dispatcher#getHead` related?

Comment (nmarton-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendProvider.scala:32]: yes they are, they are also very close (if not the same), also CacheIndex in mutable-state-cache is similar. we already have a task to understand/refactor/simplify if possible in this area.
as a first iteration we add this as an independent extension

File: ledger/participant-integration-api/src/main/scala/platform/store/IndexMetadata.scala [hunk 2]
```diff
@@ -59,6 +60,7 @@ object IndexMetadata {
       // and this property is not needed for the used ReadDao.
       participantId = Ref.ParticipantId.assertFromString("1"),
       errorFactories = errorFactories,
+      ledgerEndCache = MutableLedgerEndCache(), // not used
```
[the remark this answers is not in the document]
Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/IndexMetadata.scala:63]: this comment meant for stay

**Review by tudor-da — APPROVED** [review]

**Pull request merged**